drivers: mipi_dbi: nxp_lcdic: add support for 8080 mode

danieldegrasse opened the pull request:
Enable support for 8080 mode on the NXP LCDIC peripheral. This support has been validated on the `rd_rw612_bga` board using `lcd_par_s035_8080` shield, with the following samples:
- `samples/drivers/display`
- `samples/modules/lvgl/demos`

Additionally, this PR was tested on the `frdm_mcxn947//cpu0` using the `lcd_par_s035_8080` shield with `samples/modules/lvgl/demos`.

~Depends on #77578, #77579, and #77584 for various driver fixes (commits from those PRs are currently included in this one)~
~Blocked by https://github.com/zephyrproject-rtos/zephyr/pull/78415 due a CI fix needed~

danieldegrasse committed:
drivers: display: st7796s: respect rgb_is_inverted in 8080 8 bit mode

Update ST7796s display driver to respect the setting of rgb_is_inverted
in 8080 8 bit mode, as it was previously not applied for this mode.

Also, simplify the logic check for 16 bit mode- the new check is
functionally equivalent, but no longer inverts both boolean values.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: mipi_dbi: nxp_lcdic: add support for 8080 mode

Enable support for 8 bit 8080 mode in the NXP LCDIC driver. Support
for programming the minimum duration of the write active/inactive signal
is also added, since this will be required to support high display
clocks in 8080 mode.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: shields: lcd_par_s035: enable reset gpio within input module

The LCD_PAR_S035 display module shares the reset pin between the display
module and the touchscreen controller. The touchscreen controller needs
the reset pin to be toggled low during boot in order to select the
expected I2C address. However, the reset sequence must occur before the
display is initialized. To enable this, set a custom INPUT driver init
priority when using this display module, so that INPUT drivers start
after MIPI DBI drivers but before DISPLAY drivers.

Since the reset sequence is now operating as expected, we no longer need
to use the alt-addr probing method. This method was previously only
working correctly because boards using this display were configuring
the display INT pin using the pin control API prior to resetting the
display, so when the display was reset it would select the alt-addr
I2C address.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: nxp: rd_rw612_bga: enable with LCD_PAR_S035 shield

Enable the rd_rw612_bga board with the LCD_PAR_S035 shield. This shield
cannot be connected directly to the board, but the connection can be
made with a set of jumper wires.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: nxp: rd_rw612_bga: add display support documentation

Add documentation about supported displays on this board, as the list is
getting rather long and each display has its own bespoke connection
requirements.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

decsny commented on drivers/display/display_st7796s.c:
```diff
@@ -118,7 +118,13 @@ static int st7796s_get_pixelfmt(const struct device *dev)
+		 * invert the reported pixel format if rgb_is_inverted"
+		 * is enabled
+		 */
+		if (!(config->madctl & ST7796S_MADCTL_BGR) !=
+		    !config->rgb_is_inverted) {
```
```suggestion
                if ((bool)(config->madctl & ST7796S_MADCTL_BGR) != config->rgb_is_inverted) {
```

expression seems redundant with the double negative, also I'm not expert on the C standard but I think the first operand might need to be a boolean, at least better safe than sorry

danieldegrasse replied:
Casting certainly won't hurt. I *think* the reason the logic was written like this initially was that the `!` is going to effectively cast the value to a boolean (at least per my understanding). I'll update to make this change, and fix the logic on the lines below to match it

decsny replied:
i see, makes sense, still think it's not as clear of a way to write it though which is more important than saving a couple characters

danieldegrasse replied:
Sure- I changed the logic to match your suggestion.

decsny commented on boards/shields/lcd_par_s035/lcd_par_s035_8080.overlay:
```diff
@@ -24,8 +24,8 @@
 	gt911_lcd_par_s035: gt911-lcd_par_s035@5d {
 		compatible = "goodix,gt911";
 		reg = <0x5d>;
-		alt-addr = <0x14>;
```
why is this deleted? might want to mention in commit message

danieldegrasse replied:
I've added a blurb in the commit message. For anyone interested, the best I can tell this was working because the FRDM-MCXN947 board (previously the only in tree board using this display) was configuring the INT pin for the display in the MIPI-DBI pincontrol node. The pin control setting was such that when the RESET pin for the display (and touch screen) controller is toggled low, the touch screen controller will select the 0x14 alternate I2C address. This workaround wasn't working on the RW612 board, so I opted to properly set initialization order, thereby making sure the GT911 driver is able to pull the reset pin low and set the INT pin properly as needed to get the GT911 IC to select the standard 0x5D I2C address.

jfischer-no commented on dts/bindings/mipi-dbi/nxp,lcdic.yaml:
```diff
@@ -23,3 +23,26 @@ properties:
     description: |
       Swap bytes while transferring on LCDIC. When set, the LCDIC will send
       the most significant byte first when using multibyte pixel formats.
+
```
Do you want to remove "Only SPI mode is currently supported."(L:6)?

danieldegrasse replied:
Yes, we should. I'll update this PR, thanks for catching that

kartben commented on boards/nxp/rd_rw612_bga/doc/index.rst:
```diff
@@ -75,6 +75,59 @@ The default configuration can be found in the defconfig file:
+
+.. zephyr-app-commands::
+   :board: rd_rw612_bga
+   :gen-args: -DSHIELD=adafruit_2_8_tft_touch_v2
```
```suggestion
   :shield: adafruit_2_8_tft_touch_v2
```

kartben commented on boards/nxp/rd_rw612_bga/doc/index.rst:
```diff
@@ -75,6 +75,59 @@ The default configuration can be found in the defconfig file:
+
+.. zephyr-app-commands::
+   :board: rd_rw612_bga
+   :gen-args: -DSHIELD=lcd_par_s035_8080
```
```suggestion
   :shield: lcd_par_s035_8080
```

kartben commented on drivers/display/display_st7796s.c:
```diff
@@ -118,7 +118,13 @@ static int st7796s_get_pixelfmt(const struct device *dev)
-		if (config->madctl & ST7796S_MADCTL_BGR) {
+		/*
+		 * Similar to the handling for other interface modes,
+		 * invert the reported pixel format if rgb_is_inverted"
```
```suggestion
		 * invert the reported pixel format if "rgb_is_inverted"
```

kartben reviewed (changes requested):
Minor doc comment -- we now have a "--shield" west arg

danieldegrasse committed:
shields: lcd_par_s035: rename nxp_flexio_lcd to zephyr_mipi_dbi_parallel

Although the parallel mipi dbi mode pinout used for the LCD PAR S035
display is specific to NXP boards, the definition of the display is
generic, and does not require NXP-specific parallel mipi dbi IP.
Therefore, rename the MIPI DBI node for this display from
`nxp_flexio_lcd` to `zephyr_mipi_dbi_parallel`. The gpio-nexus node name
is unchanged, as the pinout it describes is specific to NXP.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
Rebased to pick up CI fixes

danieldegrasse commented:
@kartben ping- could you take another look at this PR?
